services/rsources/handler_test.go

```diff
@@ -5,325 +5,486 @@ import (
 	"database/sql"
 	"errors"
 	"fmt"
-	"log"
-	"os"
+	"reflect"
 	"sort"
 	"strings"
 	"sync"
-	"testing"
 	"time"
 
 	"github.com/gofrs/uuid"
 	_ "github.com/lib/pq"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
 	"github.com/ory/dockertest/v3"
-	"github.com/stretchr/testify/require"
+	"github.com/ory/dockertest/v3/docker"
 )
 
-var (
-	DB_DSN = "root@tcp(127.0.0.1:3306)/service"
-	db     *sql.DB
-)
-
-func TestMain(m *testing.M) {
-	// uses a sensible default on windows (tcp/http) and linux/osx (socket)
-	pool, err := dockertest.NewPool("")
-	if err != nil {
-		log.Fatalf("Could not connect to docker: %s", err)
-	}
-
-	database := "jobsdb"
-	// pulls an image, creates a container based on it and runs it
-	resourcePostgres, err := pool.Run("postgres", "11-alpine", []string{
-		"POSTGRES_PASSWORD=password",
-		"POSTGRES_DB=" + database,
-		"POSTGRES_USER=rudder",
-	})
-	if err != nil {
-		log.Fatalf("Could not start resource: %s", err)
-	}
-
-	port := resourcePostgres.GetPort("5432/tcp")
-	DB_DSN = fmt.Sprintf("postgres://rudder:password@localhost:%s/%s?sslmode=disable", port, database)
-	fmt.Println("DB_DSN:", DB_DSN)
-
-	// exponential backoff-retry, because the application in the container might not be ready to accept connections yet
-	if err := pool.Retry(func() error {
-		var err error
-		db, err = sql.Open("postgres", fmt.Sprintf(
-			"host=localhost port=%s user=rudder password=password dbname=jobsdb sslmode=disable",
-			port))
-		if err != nil {
-			log.Println(err)
-			return err
-		}
-		return db.Ping()
-	}); err != nil {
-		log.Fatalf("Could not connect to docker after backoff: %s", err)
-	}
-
-	code := m.Run()
-	// blockOnHold()
-	defer os.Exit(code)
-	defer func() {
-		if err := pool.Purge(resourcePostgres); err != nil {
-			log.Printf("Could not purge resource: %s \n", err)
-		}
-	}()
+type postgresResource struct {
+	resource    *dockertest.Resource
+	db          *sql.DB
+	internalDSN string
+	externalDSN string
 }
 
-func TestSourcesHandler(t *testing.T) {
-	ctx := context.Background()
-	stats := Stats{
-		In:     10,
-		Out:    4,
-		Failed: 6,
-	}
+var defaultJobTargetKey = JobTargetKey{
+	TaskRunID:     "task_run_id",
+	SourceID:      "source_id",
+	DestinationID: "destination_id",
+}
 
-	prepareService := func() (JobService, string) {
-		jobRunId := strings.ReplaceAll(uuid.Must(uuid.NewV4()).String(), "-", "")
-		sh, err := NewJobService(db)
-		require.NoError(t, err, "it should be able to create the service")
+var _ = Describe("Using sources handler", func() {
 
-		key := JobTargetKey{
-			SourceID:      "source_id",
-			DestinationID: "destination_id",
-			TaskRunID:     "task_run_id",
+	Context("standard setup with a single local datasource", Ordered, func() {
+		var (
+			pool     *dockertest.Pool
+			resource postgresResource
+			sh       JobService
+		)
+		stats := Stats{
+			In:     10,
+			Out:    4,
+			Failed: 6,
 		}
+		BeforeAll(func() {
```

Comment:
Ginkgo provides an elegant alternative to `TestMain`, only starting the relevant containers in the appropriate context.

@fracasula @lvrach

Reply:
I'm not a big fan of gingko but I'm OK with it and using contexts definitely solves this problem. As discussed [here](https://github.com/rudderlabs/rudder-server/pull/2053#issuecomment-1156131468) we also have the option of doing subtests and sometimes build tags can help too :+1: